aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,389 @@
+            }
+        }
+        self.algorithm == other.algorithm &&
+        self.priv_key == other.priv_key &&
+        self_privkey_bytes_from_ctx == other_privkey_bytes_from_ctx
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a KemPrivateKey or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPublicKey {
+    algorithm: KemAlgorithm,
+    context: LcPtr<*mut EVP_PKEY>,
+    pub_key: Vec<u8>,
+}
+
+impl KemPublicKey {
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    /// 
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    /// 
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    /// 
+    pub fn encapsulate<F, R>(&self, kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8], &[u8]) -> Result<R, Unspecified> {
+        unsafe {
+            let ctx = DetachableLcPtr::new(EVP_PKEY_CTX_new(*self.context, null_mut()))?;
+            let mut ciphertext_len;
+            let mut shared_secret_len;
+            match self.algorithm {
+                KemAlgorithm::KYBER512_R3 => {
+                    ciphertext_len = KYBER512_CIPHERTEXTBYTES;
+                    shared_secret_len = KYBER512_BYTES;
+                }
+            }
+            
+            let mut ciphertext: Vec<u8> = vec![0u8; ciphertext_len];
+            let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+            if EVP_PKEY_encapsulate(*ctx, ciphertext.as_mut_ptr(), &mut ciphertext_len,
+                                    shared_secret.as_mut_ptr(), &mut shared_secret_len) != 1 {
+                ciphertext.zeroize();
+                shared_secret.zeroize();
+                return Err(Unspecified);
+            }
+
+            #[cfg(feature = "debug-kem")]
+            {   
+                println!("Ciphertext length: {}", ciphertext_len);
+                println!("Shared secret length: {}", shared_secret_len);
+                println!("Ciphertext: {:02x?}", ciphertext);
+                println!("Shared Secret: {:02x?}", shared_secret);
+            }
+
+            kdf(&ciphertext, &shared_secret)
+        }
+    }
+
+    /// Creates a new KEM public key from raw bytes
+    /// 
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPublicKey`
+    /// 
+    /// `bytes` is a slice of raw bytes representing a `KemPublicKey`
+    /// 
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    /// 
+    pub fn from_raw_bytes(alg: KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        unsafe {
+            let pubkey_ctx = EVP_PKEY_kem_new_raw_public_key(alg.nid(), bytes.as_ptr(), alg.get_pubkey_len());
```

Comment:
Ack

Reply:
After removing macros [per Andrew's comment](https://github.com/aws/aws-lc-rs/pull/174#discussion_r1248167658) we can no longer verify that bytes.len() matches the provided algorithm. This is because in order to get the public key length we need to already have an EVP_PKEY, which is what we are creating here.